### PR TITLE
[SDK] Add AdaptingCircularBufferCounter for exponential histograms

### PR DESCRIPTION
**Pull request by euroelessar**
Updates #1391 

## Changes

Add AdaptingCircularBufferCounter, which will be used for buckets in exponential histograms.
This PR is mostly a rewrite of corresponding opentelemetry-java implementation and tests.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

File: sdk/src/metrics/data/circular_buffer.cc [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    return backing.size();
+  }
+};
+
+struct AdaptingIntegerArrayClear
+{
+  template <typename T>
+  void operator()(std::vector<T> &backing)
+  {
+    std::fill(backing.begin(), backing.end(), 0);
```
Comment (lalitb) [inline sdk/src/metrics/data/circular_buffer.cc:58]: Not sure, but you may want to typecast to `static_cast<T>(0)` to get rid of CI failure ?

**Review by lalitb — APPROVED** [review]
Thanks for the contribution. Nicely done.

**Review by esigo — APPROVED** [review]
LGTM
thanks for the PR :)

**Review by marcalff — APPROVED** [review]
LGTM.

About the CI failure:

* check CI build `CI / CMake msvc (maintainer mode) (pull_request)`
* check the logs, look for `circular_buffer.cc`, look for `error C2220`
* see suggested fix from Lalit

Thanks for the PR.

**Comment by euroelessar** [comment]
Thanks for the review.
I've been looking at coming usage of these containers in histograms themselves and had to make couple of adjustments:
1. changed index to be signed (as allowed by the spec)
2. added moving constructors/operators (to allow making histogram data point movable)

**Pull request merged**